src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6741,14 +6741,30 @@ private NameSyntax ParseQualifiedNameRight(
             }
         }
 
-        private NameSyntax RecoverFromDotDot(NameSyntax left, ref SyntaxToken separator)
+        /// <param name="separator">Passed in as the <c>..</c> token we encountered when we were consuming `name..`.
+        /// Passed out as the <c>.</c> token corresponding to the second dot in the sequence.</param>
+        private QualifiedNameSyntax RecoverFromDotDot(NameSyntax name, ref SyntaxToken separator)
         {
             Debug.Assert(separator.Kind == SyntaxKind.DotDotToken);
 
-            var leftDot = SyntaxFactory.Token(separator.LeadingTrivia.Node, SyntaxKind.DotToken, null);
-            var missingName = this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_IdentifierExpected);
-            separator = SyntaxFactory.Token(null, SyntaxKind.DotToken, separator.TrailingTrivia.Node);
-            return _syntaxFactory.QualifiedName(left, leftDot, missingName);
+            // Synthesize a dot token corresponding to the first dot in the `..` token passed in.  Place an error on it
+            // stating that there is a missing identifier immediately after it.  Importantly, place the error diagnostic
+            // on the dot token itself.  This will make this token not reusable in incremental scenarios.  The missing
+            // identifier is already not reusable (as it has zero length).  However, this does not taint the tokens
+            // around it.  We want that tainting so that an incremental reparse goes and resynthesizes the original `..`
+            // token.
```

Comment:
this is the critical part of the fix.  previously we had `dot token - missing identifier - dot token` and the incremental parser was reusing those dot tokens, even though they are not what the lexer would have produced.